### PR TITLE
fix(k8s): forward dashboards

### LINKED ISSUE #87: Missing relation in the charm based on the INTEGRATION documentation
### Bug Description

The relation is present in the documentation of INTEGRATION but there is no implementation in the charm (yet).
It would be awesome to have it done or marked in the documentation that it is expected to start version xx.yy

### To Reproduce

1. Install microk8s and install cos-lite on top of it
2. Install Charmed Kubernetes separately 
3. Create cross model relation between grafana-agent-k8s and grafana-k8s

### Environment

I ran it using Mk8s on LXD with COS lite and Charmed Kubeflow on bare-metal. The grafana-agent-k8s charm was deployed on CK8s. 

### Relevant log output

```shell
$ juju add-relation admin/cos-kubeflow.grafana-dashboards grafana-agent-k8s:grafana-dashboard
ERROR application "grafana-agent-k8s" has no "grafana-dashboard" relation

$ juju status grafana-agent-k8s
Model     Controller        Cloud/Region  Version  SLA          Timestamp
kubeflow  foundations-maas  ck8s/default  2.9.37   unsupported  07:49:56Z

SAAS                             Status  Store             URL
grafana-dashboards               active  foundations-maas  admin/cos-kubeflow.grafana-dashboards
prometheus-receive-remote-write  active  foundations-maas  admin/cos-kubeflow.prometheus-receive-remote-write

App                Version  Status  Scale  Charm              Channel  Rev  Address         Exposed  Message
grafana-agent-k8s  0.26.1   active      1  grafana-agent-k8s  stable    24  122.33.123.123  no       

Unit                  Workload  Agent  Address         Ports  Message
grafana-agent-k8s/0*  active    idle   123.123.123.123
```

### Additional context

_No response_

### LINKED ISSUE #240: No grafana dashboard on WEB if postgresql:grafana-dashboard related with grafana-agent-k8s
### Bug Description

Hi,

@simskij hints us to use `grafana-agent-k8s` [here](https://discourse.charmhub.io/t/charmed-mysql-k8s-how-to-enable-monitoring/9981/2).
Unfortunately it doesn't work out of the box: all relations established without any errors, but no grafana dashboards shown on Grafana WEB.

During the private discussion, Simmon said:
```
I remember when we tested that the last time we ended up saying "go with direct grafana dashboard relation for now" - did we not? But we definitely need to fix that asap.
```

I have tried to relate dashboard directly, but then no metrics shown on WEB.
I believe we need to trace this together to fix it and switch to `grafana-agent-k8s` in DB manuals. Tnx!

### To Reproduce

Deploy COS:
```
juju add-model cos && juju deploy cos-lite --trust
juju offer grafana:grafana-dashboard
juju offer loki:logging loki-logging
juju offer prometheus:receive-remote-write
```

Deploy DB:
```
juju add-model postgresql && juju deploy postgresql-k8s --trust --channel 14/edge -n 3
juju deploy grafana-agent-k8s agent --trust
juju consume k8s:admin/cos.grafana
juju consume k8s:admin/cos.loki
juju consume k8s:admin/cos.prometheus
```

Failed case using `grafana-agent-k8s` :
```
juju relate agent grafana-dashboards
juju relate agent loki-logging
juju relate agent prometheus
juju relate postgresql-k8s:grafana-dashboard agent
juju relate postgresql-k8s:logging agent
juju relate postgresql-k8s:metrics-endpoint agent
```
Actual results: no DB dashboards on Grafana WEB, no any errors in logs.

Try to re-relate directly:
```
juju remove-relation postgresql-k8s:grafana-dashboard agent
juju relate postgresql-k8s:grafana-dashboard grafana-dashboards
```
Actual results: dashboards appeared, but no metrics shown on dashboards.

P.S. all works well, if `postgresql-k8s` related directly with SAAS (without `grafana-agent-k8s`, as described in [manual](https://charmhub.io/postgresql-k8s/docs/h-enable-monitoring)).

### Environment

juju 3.1.5, microk8s v1.28.0
cos:
```
App           Version  Status  Scale  Charm             Channel  Rev  Address         Exposed  Message
alertmanager  0.25.0   active      1  alertmanager-k8s  stable    76  10.152.183.247  no       
catalogue              active      1  catalogue-k8s     stable    13  10.152.183.166  no       
grafana       9.2.1    active      1  grafana-k8s       stable    81  10.152.183.60   no       
loki          2.7.4    active      1  loki-k8s          stable    89  10.152.183.89   no       
prometheus    2.33.5   active      1  prometheus-k8s    stable   103  10.152.183.87   no       
traefik       2.9.6    active      1  traefik-k8s       stable   110  10.76.203.19    no  
```

db: 
```
App             Version  Status  Scale  Charm              Channel  Rev  Address         Exposed  Message
agent           0.32.1   active      1  grafana-agent-k8s  stable    38  10.152.183.127  no       
postgresql-k8s  14.9     active      3  postgresql-k8s     14/edge  128  10.152.183.111  no 
```

### Relevant log output

```shell
No errors in log.
```

### Additional context

Tnx!

### PR DESCRIPTION
**Pull request by simskij**
resolves #240 
resolves #87 

## Issue

currently, dashboards are only forwarded for the machine flavor of grafana agent

## Solution

forward dashboards for k8s as well

## Testing Instructions

1. deploy cos lite
2. deploy this bundle, with a local build of the charm:
```
bundle: pr-241
saas:
  grafana:
    url: uk8s:admin/cos.grafana
  prometheus:
    url: uk8s:admin/cos.prometheus
applications:
  ga:
    charm: local:grafana-agent-k8s-4
    channel: edge
    resources:
      agent-image: 30
    scale: 1
  pgsql:
    charm: postgresql-k8s
    channel: 14/edge
    revision: 128
    resources:
      postgresql-image: 116
    scale: 1
 relations:
- - ga:metrics-endpoint
  - pgsql:metrics-endpoint
- - ga:logging-provider
  - pgsql:logging
- - ga:send-remote-write
  - prometheus:receive-remote-write
- - ga:grafana-dashboards-provider
  - grafana:grafana-dashboard
- - ga:grafana-dashboards-consumer
  - pgsql:grafana-dashboard
```
3. look in grafana and make sure you have all the dashboards
image

## Release Notes

Forward dashboards on K8s.

### CODE REVIEW

File: src/grafana_agent.py [hunk 1]
```diff
@@ -368,7 +368,7 @@ def update_dashboards(
         try:
             dashboards = dashboards_func
         except NotImplementedError:
-            logger.debug("Dashboard forwarding is not yet enabled for k8s grafana-agent")
+            logger.error("Dashboard forwarding is not yet enabled for this substrate!")
```
Comment (dstathis) [inline src/grafana_agent.py:371]: This feels unnecessary. Let's just get rid of it.

File: src/k8s_charm.py [hunk 2]
```diff
@@ -111,6 +127,28 @@ def metrics_rules(self) -> Dict[str, Any]:
...
+                entry = {
+                    "charm": dashboards["templates"][template].get("charm", "charm_name"),
+                    "relation_id": rel.id,
+                    "title": template,
+                    "content": content,
+                }
+                aggregate[template] = entry
+
+        return aggregate.values()
+
```
Comment (dstathis) [inline src/k8s_charm.py:151]: Why aren't we just using the grafana-dashboard library for this? Seems like we are reinventing the wheel here. Also, if we do need to do this, we should put the GrafanaDashboard class in cosl. I don't like using the cos-agent lib for non cos-agent stuff.

Comment (simskij) [reply src/k8s_charm.py:151]: We need to do this for multiple reasons. The `GrafanaDashboardConsumer` works on a peer relation specific for Grafana and does a whole lot of completely unnecessary shenanigans when all we want to do is take data from N providers and give it to one consumer.

File: src/grafana_agent.py (hunk 1 above)
Comment (simskij) [reply src/grafana_agent.py:371]: ack. removed

File: src/grafana_agent.py [hunk 3]
```diff
@@ -362,14 +362,10 @@ def update_alerts_rules(
         reload_func()
 
     def update_dashboards(
-        self, dashboards_func: Any, reload_func: Callable, mapping: RulesMapping
+        self, dashboards: Any, reload_func: Callable, mapping: RulesMapping
     ) -> None:
         """Copy dashboards from relations, save them to disk, and update."""
-        try:
-            dashboards = dashboards_func
```
[the remark this answers is not in the document]
Comment (simskij) [reply src/grafana_agent.py:369]: Dylan did a partial fix, which only applied to the machine charm.

**Pull request merged**